### PR TITLE
[SPARK-40109][SQL] New SQL function: get()

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Introduce a new SQL function `get()`: Returns element of array at given (0-based) index. If the index points
     outside of the array boundaries, then this function returns NULL.
    Examples:
```
> SELECT _FUNC_(array(1, 2, 3), 2);
  2
> SELECT _FUNC_(array(1, 2, 3), 3);
  NULL
> SELECT _FUNC_(array(1, 2, 3), -1);
  NULL
```
### Why are the changes needed?

Currently, when accessing array element with invalid index under ANSI SQL mode, the error is like:
```
[INVALID_ARRAY_INDEX] The index -1 is out of bounds. The array has 3 elements. Use `try_element_at` and increase the array index by 1(the starting array index is 1 for `try_element_at`) to tolerate accessing element at invalid index and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error.
```
The provided solution is complicated. I suggest introducing a new method `get()` which always returns null on an invalid array index. This is from https://docs.snowflake.com/en/sql-reference/functions/get.html.

Note: since Spark's map access always returns null, let's don't support map type in the get method for now.

### Does this PR introduce _any_ user-facing change?

Yes, a new SQL function `get()`: returns element of array at given (0-based) index. If the index points
     outside of the array boundaries, then this function returns NULL.

### How was this patch tested?

New UT

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @srielau as well

**Review by HyukjinKwon — APPROVED** [review]
LGTM. cc @zhengruifeng and @itholic FYI

File: sql/core/src/test/resources/sql-tests/inputs/array.sql [hunk 1]
```diff
@@ -113,3 +113,9 @@ select array_size(array(true));
 select array_size(array(2, 1));
 select array_size(NULL);
 select array_size(map('a', 1, 'b', 2));
+
+-- function get()
+select get(array(1, 2, 3), 0);
+select get(array(1, 2, 3), 3);
+select get(array(1, 2, 3), null);
+select get(array(1, 2, 3), -1);
```
Comment (zhengruifeng) [inline sql/core/src/test/resources/sql-tests/inputs/array.sql:121]:
```suggestion
select get(array(1, 2, 3), -1);

```

**Review by zhengruifeng — APPROVED** [review]
the function name `get` maybe too generic, what about `array_get`?

**Comment by HyukjinKwon** [comment]
I think it's more about feature parity so `get` as is should be fine ..

**Comment by srielau** [comment]
Correct. The name get() is given.

Sent from my iPhone

On Aug 16, 2022, at 10:12 PM, Hyukjin Kwon ***@***.***> wrote:

﻿

I think it's more about feature parity so get as is should be fine ..

**Comment by gengliangwang** [comment]
@HyukjinKwon @srielau @zhengruifeng Thanks for the review. Merging to master.

**Comment by zhengruifeng** [comment]
@itholic could you please add `get` to dataframe functions when you find some time?

**Comment by zhengruifeng** [comment]
@itholic could you please add `get` to dataframe functions when you find some time? Thanks

**Pull request closed without merging**